app/drivers/tools/repair/java/SelfAPR.py

```diff
@@ -0,0 +1,122 @@
+            timeout_h,
+            self.bug_name.replace("-", "_"),
+            join(self.dir_expr, "src"),
+            join(file),
+            bug_info[self.key_fix_lines][0],
+            50, #top_n_patches
+            join(self.dir_output)
+        )        
+        status = self.run_command(
+            repair_command, self.log_output_path, "/repair/SelfAPR"
+        )
+
+        repair_command = "bash -c 'export PATH=$PATH:/root/defects4j/framework/bin && timeout -k 5m {}h python3 main.py --task validate --bug_id {} --src_dir {} --buggy_file {} --buggy_loc {} --output_folder {}'".format(
+            timeout_h,
+            self.bug_name.replace("-", "_"),
+            join(self.dir_expr, "src"),
+            join(file),
+            bug_info[self.key_fix_lines][0],
+            join(self.dir_output)
+        )
+
+        status = self.run_command(
+            repair_command,
+            self.log_output_path,
+            "/repair/SelfAPR",
+        )
+
+        self.process_status(status)
+
+        self.timestamp_log_end()
+        self.emit_highlight("log file: {0}".format(self.log_output_path))
+
+    def save_artifacts(self, dir_info):
+        """
+        Save useful artifacts from the repair execution
+        output folder -> self.dir_output
+        logs folder -> self.dir_logs
+        The parent method should be invoked at last to archive the results
+        """
+        super().save_artifacts(dir_info)
+
+    def analyse_output(self, dir_info, bug_id, fail_list):
+        """
+        analyse tool output and collect information
+        output of the tool is logged at self.log_output_path
+        information required to be extracted are:
+
+            self.stats.patches_stats.non_compilable
+            self.stats.patches_stats.plausible
+            self.stats.patches_stats.size
+            self.stats.patches_stats.enumerations
+            self.stats.patches_stats.generated
+
+            self.stats.time_stats.total_validation
+            self.stats.time_stats.total_build
+            self.stats.time_stats.timestamp_compilation
+            self.stats.time_stats.timestamp_validation
+            self.stats.time_stats.timestamp_plausible
+        """
+        
+        # extract information from output log
+        if not self.log_output_path or not self.is_file(self.log_output_path):
+            self.emit_warning("no output log file found")
+            return self.stats
+
+        self.emit_highlight(f"output log file: {self.log_output_path}")
+
+        if self.is_file(self.log_output_path):
+            log_lines = self.read_file(self.log_output_path, encoding="iso-8859-1")
+            self.stats.time_stats.timestamp_start = log_lines[0].replace("\n", "").replace("Start Time: ", "")
+            self.stats.time_stats.timestamp_end = log_lines[-1].replace("\n", "").replace("End Time: ", "")
+                
+            if not self.stats.error_stats.is_error:
+                #TODO: Implement later
```

Comment:
This got me confused at the start :sweat_smile:

Reply:
I am a little bit confused with the output path of Cerberus so I failed on several trials of parsing the results. I think I need to understand Cerberus more before implementing this part. As I have other priorities, I will do it later.

Reply:
It would be great if you could let me know how I can load the output dir at the function. I have tried to use dir_output but somehow I face the error "No such file in directory".

Reply:
No worries, I can help for that. I was just confused when I was looking at the ouput :) I can help with that

Reply:
If you are confused with output, I can explain a bit. 

- `raw_results.csv` contains all patches generated by SelfAPR. Not that the order is not ranking as SelfAPR runs 10 different models and collects all the results. It also creates many duplicated patches. 
- `validated_patches` contains the runtime information from test validation. [FE] indicates failure errors, [CE] indicates compilation errors, [Plausible] indicates that the patch passed all the test cases
- `patches` contain all the plausible patches

Note that, currently, SelfAPR provides a lot of duplicated patches as they are running in 10 different models as mentioned before.

Reply:
I see, I will play around with it then, we can push it in and then I can do commits in dev